lib/nvme/nvme_pcie.c

```diff
@@ -1607,5 +1613,65 @@ nvme_pcie_qpair_process_completions(struct spdk_nvme_qpair *qpair, uint32_t max_
 		nvme_pcie_qpair_complete_pending_admin_request(qpair);
 	}
 
+	if (qpair->ctrlr->state == NVME_CTRLR_STATE_READY) {
+
+		nvme_pcie_qpair_check_timeout(qpair);
+	}
+
 	return num_completions;
 }
+
+void
+nvme_pcie_qpair_check_timeout(struct spdk_nvme_qpair *qpair)
+{
+	uint64_t t02;
+	struct nvme_tracker *tr;
+	struct nvme_request *req;
+	struct nvme_pcie_qpair  *pqpair = nvme_pcie_qpair(qpair);
+
+	if (!TAILQ_EMPTY(&pqpair->outstanding_tr)) {
+
+		/*
+		 * qpair could be either for normal i/o or for admin command. If qpair is admin
+		 * and request is SPDK_NVME_OPC_ASYNC_EVENT_REQUEST, skip to next previous.
+		 */
+
+		tr = TAILQ_LAST(&pqpair->outstanding_tr, nvme_outstanding_tr_head);
+		while (tr->req->cmd.opc == SPDK_NVME_OPC_ASYNC_EVENT_REQUEST) {
+
+			/* qpair is for admin request */
+
+			tr = TAILQ_PREV(tr, nvme_outstanding_tr_head, tq_list);
+			if (!tr) {
+				/*
+				 * All request were AER
+				 */
+				return;
+			}
+		}
+
+		req = tr->req;
+		t02 = rte_get_timer_cycles();
+		if (req->t0 <= t02) {
+
+			if (!req->timeout_count) {
+				/*
+				 * Request has timed out. This could be i/o or admin request.
+				 * Call the registered timeout function for user to take action.
+				 */
+				req->t0 = rte_get_timer_cycles() + (rte_get_timer_hz() * NVME_IO_TIMEOUT);
+				req->timeout_count++;
+
+				if ((qpair->ctrlr != NULL) && (qpair->ctrlr->timeout_cb_fn != NULL)) {
```

Comment:
A qpair should always have a ctrlr, so I don't think we need to check for that case.